### PR TITLE
Get 0.29.x working on Python 3.11

### PR DESCRIPTION
**Pull request by da-woods**
This is a cherry pick of https://github.com/cython/cython/commit/9d1ffd5dcdcc0f2f719f86a9a30ce08a67641f10 and https://github.com/cython/cython/commit/3f56484704742c1af33c9737642125c55b0634d5 for Cython 0.29.x. It is not a hugely clean cherry-pick because `__Pyx_PyCode_New` has a different number of arguments in the two versions.

It also disables `CYTHON_FAST_PYCALL` on Python 3.11 (which I imagine is a performance regression). This is because the frame object no longer has a `f_localsplus` field to use. On Cython 3 this is fine - we use `CYTHON_VECTORCALL` instead, however that doesn't exist on 0.29.x. Therefore this is the least intrusive change available I think.

I've tested it to the extent of building Cython with these changes with Python versions 2.7, 3.7, 3.8, 3.9, 3.10, and 3.11 (twice, so it builds, then runs the compiled cython to rebuild itself) but haven't run anything more of the test-suite.

### CODE REVIEW

**Comment by da-woods** [comment]
> It also disables `CYTHON_FAST_PYCALL` on Python 3.11 (which I imagine is a performance regression). This is because the frame object no longer has a `f_localsplus` field to use. On Cython 3 this is fine - we use `CYTHON_VECTORCALL` instead, however that doesn't exist on 0.29.x. Therefore this is the least intrusive change available I think.

As an aside - this might be a good reason to target a release of Cython 3.0 sooner rather than later - it sort of feels like Cython the features needed to support modern versions of Python are in the Cython 3alpha only and it's beginning to diverge too much to easily maintain both.

**Comment by da-woods** [comment]
> It also disables `CYTHON_FAST_PYCALL` on Python 3.11 (which I imagine is a performance regression). This is because the frame object no longer has a `f_localsplus` field to use. On Cython 3 this is fine - we use `CYTHON_VECTORCALL` instead, however that doesn't exist on 0.29.x. Therefore this is the least intrusive change available I think.

Finally comment for now: It's probably possible to dedefine `__Pyx_PyFrame_GetLocalsplus` to go `frame->interpreter_frame->localsplus` and get the working again. But that's becoming a more significant change.

**Comment by scoder** [comment]
> It also disables `CYTHON_FAST_PYCALL` on Python 3.11 (which I imagine is a performance regression).

That's fine. Users who really want Py3.11 support (and all of the other goodies) can switch to Cython 3.0 now. Everyone who chooses not to do that can certainly live with a reduced feature set, since that's specifically what they choose.

File: Cython/Utility/Coroutine.c [hunk 1]
```diff
@@ -714,9 +714,14 @@ PyObject *__Pyx_Coroutine_SendEx(__pyx_CoroutineObject *self, PyObject *value, i
             PyTracebackObject *tb = (PyTracebackObject *) exc_state->exc_traceback;
             PyFrameObject *f = tb->tb_frame;
 
-            Py_XINCREF(tstate->frame);
             assert(f->f_back == NULL);
+            #if PY_VERSION_HEX >= 0x030B00A1
+            // FIXME: unclear what to do if PyThreadState_GetFrame fails
+            f->f_back = PyThreadState_GetFrame(tstate);
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:720]: [The docs](https://docs.python.org/3/c-api/init.html#c.PyThreadState_GetFrame) say that it returns `NULL` is there is no current frame. That seems reasonable and not a failure.

Comment (da-woods) [reply Cython/Utility/Coroutine.c]: I've updated the comment then. That should probably be updated in master too

**Comment by scoder** [comment]
Closing and reopening to get a CI run.

**Comment by scoder** [comment]
Not complete, but good enough for a merge. Thanks!

**Comment by vstinner** [comment]
Thanks for this cool fix!

I found this PR while trying to write a fix for `include "longintrepr.h"` while building lxml on Python 3.11. I wrote a fix and then I noticed that it's already fixed by this PR :-)

**Would it be possible to have a Cython 0.29.x release including this fix?**

It would help me to test my other changes to port projects to Python 3.11, like https://github.com/python/mypy/pull/11652 : running the mypy test suite requires to install lxml which uses Cython ;-)

**Comment by scoder** [comment]
Yeah, I was planning for new releases a few times but ran into various issues each time I tried. 0.29 finally has CI support again, so it's much safer now to try a release again. I'll give it another try soon.

**Comment by vstinner** [comment]
Oh obvisouly, please tell me if I can help you ;-)

**Comment by scoder** [comment]
FYI, 0.29.25 is out.

**Comment by vstinner** [comment]
Yeah, that's great! Thanks a lot @scoder! It's great to have Cython compatible with alpha versions of Python, it helps me a lot for testing.

**Pull request merged**